packages/near-membrane-base/src/membrane.ts

```diff
@@ -0,0 +1,1549 @@
+         * across membrane access).
+         */
+
+        /**
+         * This trap cannot just use `Reflect.has` or the `in` operator directly because
+         * the red object graph might have mutations that are only visible on the red side,
+         * which means looking into `target` directly is not viable. Instead, we need to
+         * implement a more crafty solution that looks into target's own properties, or
+         * in the red proto chain when needed.
+         *
+         * In a transparent membrane, this method will have been a lot simpler, like:
+         *
+         *      const { targetPointer } = this;
+         *      return foreignCallableHas(targetPointer, key);
+         *
+         */
+        private static dynamicHasTrap(
+            this: BoundaryProxyHandler,
+            _shadowTarget: ShadowTarget,
+            key: PropertyKey
+        ): boolean {
+            // assert: trapMutations must be true
+            const { targetPointer } = this;
+            if (foreignCallableHasOwnProperty(targetPointer, key)) {
+                return true;
+            }
+            // avoiding calling the has trap for any proto chain operation, instead we
+            // implement the regular logic here in this trap.
+            const protoPointer = foreignCallableGetPrototypeOf(targetPointer);
+            if (protoPointer === null) {
+                return false;
+            }
+            let O: object | null = getLocalValue(protoPointer);
+            // return has(O, key);
+            while (O !== null) {
+                if (ObjectProtoHasOwnProperty.call(O, key)) {
+                    return true;
+                }
+                O = getPrototypeOf(O);
+            }
+            return false;
+        }
+
+        /**
+         * This trap cannot just use `Reflect.get` directly on the `target` because
+         * the red object graph might have mutations that are only visible on the red side,
+         * which means looking into `target` directly is not viable. Instead, we need to
+         * implement a more crafty solution that looks into target's own properties, or
+         * in the red proto chain when needed.
+         *
+         * In a transparent membrane, this method will have been a lot simpler, like:
+         *
+         *   const { targetPointer } = this;
+         *   const receiverPointer = getValueOrPointer(receiver);
+         *   const foreignValueOrCallable = foreignCallableGet(targetPointer, key, receiverPointer);
+         *   return getLocalValue(foreignValueOrCallable);
+         *
+         */
+        private static dynamicGetTrap(
+            this: BoundaryProxyHandler,
+            _shadowTarget: ShadowTarget,
+            key: PropertyKey,
+            receiver: any
+        ): any {
+            // assert: trapMutations must be true
+            let O: object | null = this.proxy;
+            while (O !== null) {
+                if (ObjectProtoHasOwnProperty.call(O, key)) {
```

Comment:
`ObjectProtoHasOwnProperty.call` needs to be turned into a helper to avoid the `.call`